presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeGracefulShutdown.java

```diff
@@ -0,0 +1,99 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.nativeworker;
+
+import com.facebook.presto.testing.QueryRunner;
+import com.facebook.presto.tests.DistributedQueryRunner;
+import org.testng.annotations.Test;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+import static org.testng.Assert.assertEquals;
+
+public class TestPrestoNativeGracefulShutdown
+{
+    @Test
+    public void testGracefulShutdown() throws Exception
+    {
+        QueryRunner queryRunner = PrestoNativeQueryRunnerUtils.createNativeQueryRunner(true);
+        DistributedQueryRunner distributedQueryRunner = (DistributedQueryRunner) queryRunner;
+        URI coordinatorUri = distributedQueryRunner.getCoordinator().getBaseUrl();
+        String baseUri = coordinatorUri.toString() + "/v1/info/state";
+
+        int responseCode = sendShutdownRequest(baseUri, "\"INVALID_SHUTDOWN\"");
+        assertEquals(responseCode, 400, "Expected a 400 Bad Request response for invalid body");
+
+        responseCode = sendShutdownRequest(baseUri, "");
+        assertEquals(responseCode, 400, "Expected a 400 Bad Request response for empty body");
+
+        responseCode = sendShutdownRequest(baseUri, "\"SHUTTING_DOWN\"");
+        assertEquals(responseCode, 200, "Expected a 200 OK response for valid shutdown request");
+        String state = new String();
+        try {
+            URL url = new URL(baseUri);
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            connection.setRequestMethod("GET");
+            connection.setDoOutput(true);
+            connection.setRequestProperty("Content-Type", "application/json");
+
+            int resCode = connection.getResponseCode();
+            if (resCode == HttpURLConnection.HTTP_OK) { // 200
+                try (BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
+                    StringBuilder response = new StringBuilder();
+                    String line;
+                    while ((line = in.readLine()) != null) {
+                        response.append(line);
+                    }
+                    state = response.toString();
+                }
+            }
+            else {
+                System.out.println("GET request failed. Response Code: " + responseCode);
+            }
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+        assertEquals(state, "\"SHUTTING_DOWN\"");
+
+        queryRunner.close();
+    }
+
+    public static int sendShutdownRequest(String shutdownUrl, String body)
+    {
+        try {
+            URL url = new URL(shutdownUrl);
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```

Comment:
DistributedQueryRunner has a HttpClient. Can we not use that?
That API seems simpler. The same is true above to check the status of the node.

Reply:
getCoordinatorInfoState() is private, so is client. Do you want me to extend DistributedQueryRunner class with setter getters?

Reply:
Thanks for the insight. The code looks much cleaner. I have one issue though I am not able to extract statusCode from JettyClient because JettyResponse is private and I need to pass a custom responseHandler to get the status code so I am using the same java.net.HttpURLConnection class that I was using in sendShutdownRequest function.